Feature/create user

guerretta16 opened the pull request:
#13 
- `username` added in model's User fillable.
-  Route file created and added into Route Service Provider.
-  Create User request created.
   - Validate the email and username as unique, password and confirm_password with regex and de name.
- Controller `CreateUser` added.
   - Doesn't have a middleware, 'cause anyone can create a user.
   - Verify if the `password` and de `confirm_password` match.
   - Password with hash. 
 
### API ENDPOINT
- Create user as `/users/`.

guerretta16 committed:
refactor(models): `username` field added to `fillable`.

guerretta16 committed:
feat(routes): `users.php` file added.

- File `users.php` added to define user's routes.
- User's route map added into `RouteServiceProvider`.

guerretta16 committed:
feat(request): Create user request file added.

- The rules to validate a new user were added.
-- `email` and `username` unique.
- Two field for password and their confirmation.
- Passwords has a regex validation that accept:
-- 8 characters at least.
-- 1 Upper letter at least.
-- 1 Lower letter at least.
-- At least one number or none.
-- Special characters such as '.@$!%*?&'.

guerretta16 committed:
feat(controllers): Controller `CreateUser` added.

- Doesn't have middleware, 'cause anyone can create a user.
- request body validated with `CreateUserRequest`.
- Validate if `password` and `confirm_password` are the same, if not, returns an error with status `400`.
- Hash the password too.

f-esquivel commented on app/Http/Controllers/Users/CreateUserController.php:
```diff
@@ -0,0 +1,37 @@
+                'username' => $data['username'],
+                'name' => $data['name'],
+                'password' => Hash::make($data['password']),
+                'type' => UserTypesConstant::USER
```
La asignación es innecesaria pq el valor por defecto es `USER`?

f-esquivel commented on app/Http/Requests/Users/CreateUserRequest.php:
```diff
@@ -0,0 +1,25 @@
+        return [
+            'email' => 'required|email|unique:users,email',
+            'username' => 'required|string|unique:users,username',
+            'password' => 'required|string|regex:/^(?=.*[a-z])(?=.*[A-Z])(?=.*[.@$!%*?&])[A-Za-z\d.@$!%*?&]{8,}$/',
+            'confirm_password' => 'required|string|regex:/^(?=.*[a-z])(?=.*[A-Z])(?=.*[.@$!%*?&])[A-Za-z\d.@$!%*?&]{8,}$/',
```
Qué casos cubre la Regex?

f-esquivel commented on app/Http/Controllers/Users/CreateUserController.php:
```diff
@@ -0,0 +1,37 @@
+    {
+        $data = $request->validated();
+
+        if ($data['password'] !== $data['confirm_password']) {
+            return response()->json([
+                'Message' => "Passwords don't match. Please verify the confirm password input."
+            ], 400);
```
Implementar `Form request rule`

guerretta16 committed:
refactor(request): `same` rule was added for `confirm_password`.

Rule added in `CreateUserRequest`.

guerretta16 committed:
refactor(controllers): Refactor in `CreateUserController`.

- Validation for password was removed and added in the form request.
- Create function changed.
- Token added in the responses as login.

guerretta16 committed:
refactor(request): Refactor in `CreateUserRequest`.

- Regex removed form `confirm_password`.

guerretta16 committed:
refactor(controller): Refactor in `CreateUserController`.

- `type` was removed from create function.

f-esquivel reviewed (approved):
🤠 👌🏾
